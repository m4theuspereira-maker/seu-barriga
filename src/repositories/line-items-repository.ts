import { PrismaClient } from "@prisma/client";
import { IRepository } from "./interfaces/interfaces";
import { ObjectId } from "mongodb";

export class LineItemsRepository implements IRepository {
  constructor(private readonly client: PrismaClient) {}

  async create(item: { name: string; price: string; ammount: number }) {
    return this.client.lineItem.create({
      data: { ...item, id: new ObjectId().toString() }
    });
  }

  async createMany(
    lineItems: { price: string; name: string; ammount: number }[]
  ) {
    const data = lineItems.map((lineItem) => ({
      ammount: lineItem.ammount,
      price: lineItem.price,
      name: lineItem.name,
      id: new ObjectId().toString()
    }));

    return this.client.lineItem.createMany({ data });
  }
  update(id: string, updatePlayload: any): Promise<any> {
    throw new Error("Method not implemented.");
  }

  async findMany(names: string[]): Promise<
    {
      price: string;
      name: string;
      ammount?: number;
      guruPaymentLink?: string | null;
    }[]
  > {
    return this.client.lineItem.findMany({ where: { name: { in: names } } });
  }

  findOne(input: any): Promise<any> {
    throw new Error("Method not implemented.");
  }
}
