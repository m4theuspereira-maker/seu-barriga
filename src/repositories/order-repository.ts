import { PrismaClient } from "@prisma/client";
import { IOrder, IOrderFound, IRepository } from "./interfaces/interfaces";
import { ObjectId } from "mongodb";

export class OrderRepository implements IRepository {
  constructor(private readonly client: PrismaClient) {}

  create(order: IOrder): Promise<any> {
    return this.client.order.create({
      data: { id: new ObjectId().toString(), ...order }
    });
  }

  async findMany(ip: string): Promise<IOrderFound[]> {
    return this.client.order.findMany({
      where: { ip },
      orderBy: { createdAt: "desc" }
    });
  }
  async findOne(ip: string): Promise<IOrderFound | null> {
    return this.client.order.findFirst({ where: { ip } });
  }
}
