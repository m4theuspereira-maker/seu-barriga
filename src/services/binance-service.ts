import {
  BINANACE_PAY_BASE_URL,
  BINANCE_API_KEY,
  BINANCE_CANCEL_URL,
  BINANCE_GET_TIME_URL,
  BINANCE_SECRET_KEY,
  BINANCE_SUCESS_URL
} from "../common/environment-consts";
import { HttpClient } from "../config/client";
import { OrderRepository } from "../repositories/order-repository";
import {
  IDeliveryInformation,
  ILineItem,
  ILineItemsService,
  IPaymentService,
  IProductLineItem
} from "./interfaces/interfaces";
import * as crypto from "crypto";
import { TelegramService } from "./telegram-service";

export class BinanceService implements IPaymentService {
  constructor(
    private readonly httpClient: HttpClient,
    private readonly orderRepository: OrderRepository,
    private readonly telegramService: TelegramService,
    private readonly lineItemsService: ILineItemsService
  ) {}

  async makeCheckout(
    lineItems: IProductLineItem[],
    ip: string,
    externalOrderId: string,
    deliveryInformation?: IDeliveryInformation | undefined
  ): Promise<{ url: string | null }> {
    let lineItemsToBeSend: ILineItem[] = [];
    const products = lineItems.map((lineItem) => lineItem.name);

    if (ip) {
      await this.orderRepository.create({ ip, externalOrderId });
    }

    lineItemsToBeSend = await this.lineItemsService.getLineItemsToBeSend(
      products,
      lineItems
    );

    const body = this.makeRequestBody(
      this.lineItemsService.calculateAmmount(lineItemsToBeSend),
      "MIROHOBA"
    );

    const {
      data: { serverTime }
    } = await this.httpClient.get(BINANCE_GET_TIME_URL);

    const nonce = this.generateNonce(32);

    const signature = this.generateSignature(
      `${serverTime}\n${nonce}\n${JSON.stringify(body)}\n`
    );

    const { data } = await this.httpClient.post(
      `${BINANACE_PAY_BASE_URL}/binancepay/openapi/v2/order`,
      body,
      { headers: this.makeHeader(serverTime, nonce, signature) as any }
    );

    if (deliveryInformation?.firstName != undefined) {
      await this.telegramService.sendMessageToTelegramBot(
        this.telegramService.formatMessage(deliveryInformation)
      );
    }

    return { url: data.data.universalUrl };
  }

  private generateNonce(length: number) {
    const characters =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    let result = "";
    const charactersLength = characters.length;
    for (let i = 0; i < length; i++) {
      result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
  }

  private makeHeader(serverTime: number, nonce: String, signature: string) {
    return {
      "content-type": "application/json",
      "BinancePay-Timestamp": serverTime, // UnixTimestamp em milissegundos
      "BinancePay-Nonce": nonce, // Substitua com o valor aleatÃ³rio desejado
      "BinancePay-Certificate-SN": BINANCE_API_KEY, // Substitua com a chave certificado
      "BinancePay-Signature": signature
    };
  }

  private generateSignature(payload: string) {
    const hmac = crypto.createHmac("sha512", BINANCE_SECRET_KEY!);
    hmac.update(payload);
    return hmac.digest("hex").toUpperCase();
  }

  private makeRequestBody(ammount: number, productName: string) {
    return {
      env: {
        terminalType: "WEB"
      },
      orderTags: {
        ifProfitSharing: false
      },
      merchantTradeNo: new Date().getTime().toString(),
      fiatAmount: ammount,
      fiatCurrency: "USD",
      returnUrl: BINANCE_SUCESS_URL,
      cancelUrl: BINANCE_CANCEL_URL,
      goods: {
        goodsType: "01",
        goodsCategory: "Z000",
        referenceGoodsId: "p0ZoB1FwH6",
        goodsName: productName,
        goodsDetail: "MIHOROBA"
      }
    };
  }

  async switchOrderStatus(ip: string, status: string): Promise<void> {
    return await this.lineItemsService.switchOrderStatus(ip, status);
  }
}
