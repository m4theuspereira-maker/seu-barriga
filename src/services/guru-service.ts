import { HttpClient } from "../config/client";
import { OrderRepository } from "../repositories/order-repository";
import {
  IDeliveryInformation,
  ILineItem,
  ILineItemsService,
  IPaymentService
} from "./interfaces/interfaces";
import { TelegramService } from "./telegram-service";

export class GuruService implements IPaymentService {
  constructor(
    private readonly orderRepository: OrderRepository,
    private readonly lineItemsService: ILineItemsService,
    private readonly telegramService: TelegramService
  ) {}

  async makeCheckout(
    lineItems: any[],
    ip: string,
    externalOrderId: string,
    deliveryInformation?: IDeliveryInformation | undefined
  ): Promise<{ url: string | null }> {
    let lineItemsToBeSend: ILineItem[] = [];
    const products = lineItems.map((lineItem) => lineItem.name);

    if (ip) {
      await this.orderRepository.create({ ip, externalOrderId });
    }

    lineItemsToBeSend = await this.lineItemsService.getLineItemsToBeSend(
      products,
      lineItems
    );

    if (deliveryInformation?.firstName != undefined) {
      await this.telegramService.sendMessageToTelegramBot(
        this.telegramService.formatMessage({ ...deliveryInformation, products })
      );
    }

    return { url: lineItemsToBeSend.at(0)?.guruPaymentLink! };
  }

  async switchOrderStatus(ip: string, status: string): Promise<void> {
    return await this.lineItemsService.switchOrderStatus(ip, status);
  }
}
