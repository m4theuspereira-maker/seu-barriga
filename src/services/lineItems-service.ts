import {
  WOOCOMMERCE_COSTUMER_KEY,
  WOOCOMMERCE_COSTUMER_SECRET
} from "../common/environment-consts";
import { HttpClient } from "../config/client";
import { LineItemsRepository } from "../repositories/line-items-repository";
import { OrderRepository } from "../repositories/order-repository";
import {
  ILineItem,
  ILineItemsService,
  IProductLineItem
} from "./interfaces/interfaces";

export class LineItemsService implements ILineItemsService {
  constructor(
    private readonly lineItemsRepository: LineItemsRepository,
    private readonly orderRepository: OrderRepository,
    private readonly httpClient: HttpClient
  ) {}

  async switchOrderStatus(ip: string, status: string): Promise<any> {
    const orderFound = (await this.orderRepository.findMany(ip)).at(0);
    const url = `https://mimosapowders.com/wp-json/wc/v3/orders/${
      orderFound?.externalOrderId
    }?consumer_key=${WOOCOMMERCE_COSTUMER_KEY!}&consumer_secret=${WOOCOMMERCE_COSTUMER_SECRET!}`;
    const result = await this.httpClient.put(
      url,
      { status },
      { headers: { "Content-Type": "application/json" } }
    );

    return result.data;
  }

  async getLineItemsToBeSend(
    products: string[],
    lineItems: IProductLineItem[]
  ) {
    let lineItemsToBeSend: ILineItem[] = [];
    const productsFound = await this.lineItemsRepository.findMany(products);

    lineItems.forEach((lineItem) => {
      const priceFound = productsFound.find(
        (priceId) =>
          priceId.name.toLowerCase().trim() ===
          lineItem.name.toLowerCase().trim()
      );

      if (priceFound?.name) {
        lineItemsToBeSend.push({
          price: priceFound?.price!,
          quantity: Number(lineItem.quantity),
          ammount: priceFound?.ammount,
          guruPaymentLink: priceFound.guruPaymentLink!
        });
      }
    });

    return lineItemsToBeSend;
  }

  calculateAmmount(lineItems: ILineItem[]) {
    return lineItems.reduce((accumulator, currentValue: ILineItem) => {
      const product = currentValue.ammount! * currentValue.quantity;
      return accumulator + product;
    }, 0);
  }
}
