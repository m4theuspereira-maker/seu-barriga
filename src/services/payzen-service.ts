import {
  PAYZEN_API_PASSWORD,
  PAYZEN_API_USERNAME,
  PAYZEN_BASE_URL
} from "../common/environment-consts";
import { HttpClient } from "../config/client";
import { OrderRepository } from "../repositories/order-repository";
import {
  IDeliveryInformation,
  ILineItem,
  ILineItemsService,
  IPaymentService
} from "./interfaces/interfaces";
import { TelegramService } from "./telegram-service";

export class PayzenService implements IPaymentService {
  constructor(
    private readonly orderRepository: OrderRepository,
    private readonly lineItemsService: ILineItemsService,
    private readonly httpClient: HttpClient,
    private readonly telegramService: TelegramService
  ) {}

  async makeCheckout(
    lineItems: any[],
    ip: string,
    externalOrderId: string,
    deliveryInformation?: IDeliveryInformation | undefined
  ): Promise<{ url: string | null }> {
    let lineItemsToBeSend: ILineItem[] = [];
    const products = lineItems.map((lineItem) => lineItem.name);

    if (ip) {
      await this.orderRepository.create({ ip, externalOrderId });
    }

    lineItemsToBeSend = await this.lineItemsService.getLineItemsToBeSend(
      products,
      lineItems
    );

    const token = Buffer.from(
      `${PAYZEN_API_USERNAME}:${PAYZEN_API_PASSWORD}`
    ).toString("base64");

    const amount = this.lineItemsService.calculateAmmount(lineItemsToBeSend);

    const {
      data: {
        USDBRL: { high }
      }
    } = await this.httpClient.get(
      "https://economia.awesomeapi.com.br/last/USD-BRL"
    );

    const amountToBePaid = amount * high;

    let additionalInformations = {};
    const body = {
      amount: Math.round(amountToBePaid * 100),
      currency: "BRL",
      locale: "en_EN",
      orderId: new Date().getTime().toString()
    } as any;

    if (deliveryInformation?.firstName != undefined) {
      await this.telegramService.sendMessageToTelegramBot(
        this.telegramService.formatMessage({ ...deliveryInformation, products })
      );

      additionalInformations = {
        firstName: deliveryInformation?.firstName,
        lastName: deliveryInformation?.lastName,
        phoneNumber: deliveryInformation?.phone,
        streetNumber: deliveryInformation?.streetHouseNumber,
        address: deliveryInformation?.additionalInformation,
        zipCode: deliveryInformation?.zipCode,
        city: deliveryInformation?.zipCode,
        state: deliveryInformation?.state,
        country: deliveryInformation?.country
      };

      body.customer = {
        email: deliveryInformation?.emailAddress,
        billingDetails: {
          category: "PRIVATE",
          ...additionalInformations
        }
      };
    }

    const { data } = await this.httpClient.post(
      `${PAYZEN_BASE_URL}/api-payment/V4/Charge/CreatePaymentOrder`,
      body,
      {
        headers: {
          Authorization: `Basic ${token}`,
          "Content-Type": "application/json"
        }
      } as any
    );

    return { url: data.answer.paymentURL };
  }

  async switchOrderStatus(ip: string, status: string): Promise<void> {
    return await this.lineItemsService.switchOrderStatus(ip, status);
  }
}
