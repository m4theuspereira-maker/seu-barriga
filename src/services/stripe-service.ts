import Stripe from "stripe";
import {
  IDeliveryInformation,
  ILineItem,
  IPaymentService,
  IProductLineItem
} from "./interfaces/interfaces";
import { LineItemsRepository } from "../repositories/line-items-repository";
import {
  STRIPE_CANCEL_URL,
  STRIPE_SUCESS_URL,
  WOOCOMMERCE_COSTUMER_KEY,
  WOOCOMMERCE_COSTUMER_SECRET
} from "../common/environment-consts";
import { OrderRepository } from "../repositories/order-repository";
import { HttpClient } from "../config/client";
import { TelegramService } from "./telegram-service";

export class StripeService implements IPaymentService {
  constructor(
    private readonly stripeConfig: Stripe,
    private readonly lineItemsRepository: LineItemsRepository,
    private readonly orderRepository: OrderRepository,
    private readonly httpClient: HttpClient,
    private readonly telegramService: TelegramService
  ) {}

  async switchOrderStatus(ip: string, status: string): Promise<void> {
    const orderFound = (await this.orderRepository.findMany(ip)).at(0);
    const url = `https://mimosapowders.com/wp-json/wc/v3/orders/${
      orderFound?.externalOrderId
    }?consumer_key=${WOOCOMMERCE_COSTUMER_KEY!}&consumer_secret=${WOOCOMMERCE_COSTUMER_SECRET!}`;
    const result = await this.httpClient.put(
      url,
      { status },
      { headers: { "Content-Type": "application/json" } }
    );
  }

  async makeCheckout(
    lineItems: Array<IProductLineItem>,
    ip: string,
    externalOrderId: string,
    deliveryInformation?: IDeliveryInformation
  ) {
    let lineItemsToBeSend: ILineItem[] = [];
    const products = lineItems.map((lineItem) => lineItem.name);
    const deliveruInformationWithProdutcts = {
      ...deliveryInformation,
      products
    };

    if (ip) {
      await this.orderRepository.create({ ip, externalOrderId });
    }

    const productsFound = await this.lineItemsRepository.findMany(products);

    lineItems.forEach((lineItem) => {
      const priceFound = productsFound.find(
        (priceId) =>
          priceId.name.toLowerCase().trim() ===
          lineItem.name.toLowerCase().trim()
      );

      if (priceFound?.name) {
        lineItemsToBeSend.push({
          price: priceFound?.price!,
          quantity: Number(lineItem.quantity)
        });
      }
    });

    return this.intergrateWithStripe(
      lineItemsToBeSend,
      deliveruInformationWithProdutcts as any
    );
  }

  private async intergrateWithStripe(
    lineItems: Array<ILineItem>,
    deliveryInformation?: IDeliveryInformation
  ) {
    const { url } = await this.stripeConfig.checkout.sessions.create({
      line_items: lineItems,
      mode: "payment",
      success_url: STRIPE_SUCESS_URL!,
      cancel_url: STRIPE_CANCEL_URL!,
      allow_promotion_codes: true
    });

    if (deliveryInformation?.firstName != undefined) {
      await this.telegramService.sendMessageToTelegramBot(
        this.telegramService.formatMessage(deliveryInformation)
      );
    }

    return { url: url };
  }
}
