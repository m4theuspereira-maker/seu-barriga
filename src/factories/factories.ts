import axios from "axios";
import { client } from "../config/client";
import { stripeConfig } from "../config/stripe-config";
import { PaymentController } from "../controllers/payment-controller";
import { LineItemsRepository } from "../repositories/line-items-repository";
import { StripeService } from "../services/stripe-service";
import { OrderRepository } from "../repositories/order-repository";
import { TelegramService } from "../services/telegram-service";
import { BinanceService } from "../services/binance-service";
import { LineItemsService } from "../services/lineItems-service";

export const telegramService = new TelegramService(axios);

export function paymentFactory() {
  return new PaymentController(
    new StripeService(
      stripeConfig,
      new LineItemsRepository(client),
      new OrderRepository(client),
      axios,
      telegramService
    )
  );
}

export function binanceFactory() {
  return new BinanceService(
    axios,
    new OrderRepository(client),
    telegramService,
    new LineItemsService(
      new LineItemsRepository(client),
      new OrderRepository(client),
      axios
    )
  );
}

export function binanceControlleFactory(binanceFactory: BinanceService) {
  return new PaymentController(binanceFactory);
}
