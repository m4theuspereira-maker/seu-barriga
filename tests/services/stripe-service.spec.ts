import { beforeAll, describe, expect, it } from "vitest";
import { StripeService } from "../../src/services/stripe-service";
import { stripeConfig } from "../../src/config/stripe-config";
import { LineItemsRepository } from "../../src/repositories/line-items-repository";
import { OrderRepository } from "../../src/repositories/order-repository";
import axios from "axios";
import { telegramService } from "../../src/factories/factories";
import { client } from "../../src/config/client";

describe("StripeService", () => {
  let stripeService: StripeService;

  beforeAll(() => {
    stripeService = new StripeService(
      stripeConfig,
      new LineItemsRepository(client),
      new OrderRepository(client),
      axios,
      telegramService
    );
  });

  describe("formatMessage", () => {
    it("should return products array as string", () => {
      const result = telegramService.formatMessage({
        firstName: "Matheus",
        lastName: "Santos Pereira",
        country: "Brasil",
        city: "Tobias Barreto",
        streetHouseNumber: "Rua Carlos Lemos, 638",
        houseType: "",
        state: "- Select Concelho -",
        zipCode: "943000000",
        phone: "9999999999999999",
        emailAddress: "matheusmonaco1@hotmail.com",
        additionalInformation: "",
        products: ["Mimosa H. Inner Root Bark (MHRB) – Powder - 0,5kg"]
      });

      expect(result).toMatchSnapshot(
        `🛍️ **Nova Compra Realizada** 🛍️

      ℹ️ **Detalhes da Compra:**
      🗣️ Para: Matheus Santos Pereira,
      📦 Produtos: Mimosa H. Inner Root Bark (MHRB) – Powder - 0,5kg
      🌎 País: Brasil
      🏡 Estado: - Select Concelho -
      🏙️ Cidade: Tobias Barreto
      📬 ZipCode: 943000000
      📞 Telefone: 9999999999999999
      📧 E-mail: matheusmonaco1@hotmail.com
      🛣️ Rua e número: Rua Carlos Lemos, 638
      🏘️ Tipo de residência:
      🗞️ Informações Adicionais: `
      );
    });
  });
});
