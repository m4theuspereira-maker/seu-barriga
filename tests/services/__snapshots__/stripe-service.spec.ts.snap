// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StripeService > formatMessage > should return products array as string > 🛍️ **Nova Compra Realizada** 🛍️

      ℹ️ **Detalhes da Compra:**
      🗣️ Para: Matheus Santos Pereira,
      📦 Produtos: Mimosa H. Inner Root Bark (MHRB) – Powder - 0,5kg
      🌎 País: Brasil
      🏡 Estado: - Select Concelho -
      🏙️ Cidade: Tobias Barreto
      📬 ZipCode: 943000000
      📞 Telefone: 9999999999999999
      📧 E-mail: matheusmonaco1@hotmail.com
      🛣️ Rua e número: Rua Carlos Lemos, 638
      🏘️ Tipo de residência:
      🗞️ Informações Adicionais:  1`] = `
"
🛍️ **Nova Compra Realizada** 🛍️

ℹ️ **Detalhes da Compra:**

🗣️ Para: Matheus Santos Pereira,
📦 Produtos: Mimosa H. Inner Root Bark (MHRB) – Powder - 0,5kg
🌎 País: Brasil
🏡 Estado: - Select Concelho -
🏙️ Cidade: Tobias Barreto
📬 ZipCode: 943000000
📞 Telefone: 9999999999999999
📧 E-mail: matheusmonaco1@hotmail.com
🛣️ Rua e número: Rua Carlos Lemos, 638
🏘️ Tipo de residência: 
🗞️ Informações Adicionais: 
    "
`;
